Create public contructors for StackExchange.Redis event args for testing purpose

n1l opened the pull request:
Create public contructors for StackExchange.Redis event args to be able to create tests scenarios based on redis events. Add sealed attribute to other EventArgs for consistency.

Issue with details - https://github.com/StackExchange/StackExchange.Redis/issues/1325

mgravell commented on src/StackExchange.Redis/HashSlotMovedEventArgs.cs:
```diff
@@ -27,7 +27,15 @@ public sealed class HashSlotMovedEventArgs : EventArgs, ICompletable
+        /// <param name="hashSlot">Hash slot.</param>
+        /// <param name="old">Old endpoint.</param>
+        /// <param name="new">New endpoint.</param>
+        public HashSlotMovedEventArgs(EventHandler<HashSlotMovedEventArgs> handler, object sender,
```
this is a very unusual API, for specific reasons; I don't think any external caller (for unit testing etc) would need this - or indeed could use this; I wonder if the `public` constructor should just be:

``` c#
public HashSlotMovedEventArgs(int hashSlot, EndPoint old, EndPoint @new) : this (null, null, hashSlot, old, new) {}
```

thoughts?

n1l replied:
Still this is a public event and someone may be interested to unit-test some scenarios.

mgravell replied:
My answer to that, then: is *show* the scenario that needs that, ideally in a test.  Adding something without a demonstrated use-case is ... not a good way to design the API usually.

n1l replied:
@mgravell this is an outdated comment. I've updated a pr.

mgravell commented on src/StackExchange.Redis/ConnectionFailedEventArgs.cs:
```diff
@@ -11,7 +11,19 @@ public sealed class ConnectionFailedEventArgs : EventArgs, ICompletable
+         /// <param name="failureType">Redis connection failure type.</param>
+         /// <param name="exception">The exception occured.</param>
+         /// <param name="physicalName">Connection physical name.</param>
+        public ConnectionFailedEventArgs(EventHandler<ConnectionFailedEventArgs> handler, object sender, EndPoint endPoint, ConnectionType connectionType, ConnectionFailureType failureType, Exception exception, string physicalName)
```
similar to the comment on `EndPointEventArgs` - I wonder if the `public` .ctor should not expose `handler` / `sender`

n1l replied:
Answer is the same.

mgravell commented on src/StackExchange.Redis/InternalErrorEventArgs.cs:
```diff
@@ -7,11 +7,21 @@ namespace StackExchange.Redis
+        /// <param name="connectionType">Redis connection type.</param>
+        /// <param name="exception">The exception occured.</param>
+        /// <param name="origin">Origin.</param>
+        public InternalErrorEventArgs(EventHandler<InternalErrorEventArgs> handler, object sender, EndPoint endpoint, ConnectionType connectionType, Exception exception, string origin)
```
ditto

mgravell commented on src/StackExchange.Redis/RedisErrorEventArgs.cs:
```diff
@@ -11,7 +11,15 @@ public sealed class RedisErrorEventArgs : EventArgs, ICompletable
+        /// <param name="sender">The source of the event.</param>
+        /// <param name="endpoint">Redis endpoint.</param>
+        /// <param name="message">Error message.</param>
+        public RedisErrorEventArgs(
```
ditto

mgravell commented:
The **very unusual** construction API here is because of the way that `ICompletable` works; I can't see any unit-test scenarios where that would be necessary/appropriate. Now, this *could* just be that I'm being unimaginative, but... I *suspect* we can drop those and leave more idiomatic APIs.

In particular, I don't want to force us to make an implementation detail part of the public API.

Thoughts and discussion welcome.

---

Also: it would **really help** to see these how they are intended to be used, i.e. in a unit test that uses them *how you want to*.

n1l commented:
Oh, mb you're talking about public ctor's without params at all? Like for testing purpose?

mgravell commented:
> Oh, mb you're talking about public ctor's without params at all? Like for testing purpose?

no, it is just the handler and sender that are atypical; the API here reverses that (because: reasons). You would not usually have those as parameters on *classic* event args. So I'm saying: if the args has those: create *new* `public` .ctors that invoke the existing `internal` .ctors, passing `null` for `handler` and `args`. If the old .ctor *only* has those parameters, then yes: that would result in a parameterless constructor.

mgravell commented:
see also: the example I gave for `HashSlotMovedEventArgs`

n1l commented:
@mgravell 
Yes I've understood the idea a bit later :)

n1l commented:
@mgravell I've updated please take a look.

n1l commented:
@mgravell  why?
Please can you tell me what's your opinion. :)

n1l commented:
@mgravell please take a look again :)

n1l commented:
@mgravell is everything ok? :)

mgravell commented:
I know you don't mean anything by it, but keep in mind that OSS maintainers aren't around 24/7 to jump on reviews - lost of plates spinning, including jobs, family, sleep, etc; I plan on taking another look as soon as I have a moment, probably later today.

n1l commented:
@mgravell  Thank you! :)

NickCraver commented:
Minor language note here, could you please change all descriptions from "The constructor for testing purpose." to "This constructor is only for testing purposes."?

n1l commented:
> This constructor is only for testing purposes."?
Shure!

mgravell commented:
Constructors look good, but there's still nothing test-wise that indicates the intended usage. What makes me nervous is that as a general principle of API design, I usually prefer to see an example usage along with any new API (including new methods/overloads); that's usually how we discover whether the proposed API actually achieves the things we want to demonstrate. Is there any way of adding a meaningful test here? Not just "oh look, I can call a constructor" - but rather: something that actually shows *why we're adding it*, and that we have achieved what we wanted. If that means adding Moq or whatever to the test project, I'm fine with that...

n1l commented:
> Is there any way of adding a meaningful test here?

Yes of course, I was just a bit lazy to add them because I thought it's obvious, but ok it's not a big deal again.

> Not just "oh look, I can call a constructor" - but rather: something that actually shows _why we're adding it_, and that we have achieved what we wanted.

Do you mean something except tests(documentation, comments samples) or only tests?

mgravell commented:
> Do you mean something except tests(documentation, comments samples) or only tests?

I'm just talking about tests here; this isn't an API that needs documentation - just: you're adding this API because you want to do a thing; it would be good to see the thing happening, so we can prove that it is the right API and enables the thing.

n1l commented:
Hi, @NickCraver , @mgravell !
I've added some tests for the feature. I've removed sealed attribute because substitute mocking framework can't create proxy for a sealed class. Hope it's ok.

So the Idea is - if I have a component which can do some diagnostic for me(logging, pushing events etc) I want to tests it's handlers. But if there is no public c'tors for event args I can't invoke such handlers and test my component.

mgravell commented:
Awesome. Will look!

On Sat, 1 Feb 2020, 16:30 n1l, <notifications@github.com> wrote:

> Hi, @NickCraver <https://github.com/NickCraver> , @mgravell
> <https://github.com/mgravell> !
> I've added some tests for the feature. I've removed sealed attribute
> because substitute mocking framework can't create proxy for a sealed class.
> Hope it's ok.
>
> So the Idea is - if I have a component which can do some diagnostic for
> me(logging, pushing events etc) I want to tests it's handlers. But if there
> is no public c'tors for event args I can't invoke such handlers and test my
> component.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/StackExchange/StackExchange.Redis/pull/1326?email_source=notifications&email_token=AAAEHMBRAOQ65H3SY6MZYQ3RAWPSHA5CNFSM4KJSCLM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKRA53I#issuecomment-581045997>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAEHMB2S772CCG6ECDBXP3RAWPSHANCNFSM4KJSCLMQ>
> .
>

mgravell commented:
Merging, thanks
